### ISSUE TITLE
Generative Predictive Coding

### ISSUE THREAD

**Issue by ToddMorrill** [Bogacz-Group/PredictiveCoding#3]
Hello - I'm using your codebase as the basis for a class project and I have some questions if you're willing to weigh in.

1. The distinction between "standard" generative PC (as presented in the [Bogacz tutorial](https://www.sciencedirect.com/science/article/pii/S0022249615000759)) and "reverse" discriminative PC (as presented in the [2017 Whittington & Bogacz paper](https://direct.mit.edu/neco/article/29/5/1229/8261/An-Approximation-of-the-Error-Backpropagation)) has recently become clearer to me. To confirm, is this codebase primarily geared towards "reverse" discriminative PC? In particular, your network inputs are data, outputs are labels/targets, feedforward activities are predictions, and feedback activities are essentially gradients (like a standard neural network). [Image source](https://arxiv.org/abs/2107.12979) <img width="1257" alt="PNG image" src="https://github.com/user-attachments/assets/7cfaed5b-afc7-4fd0-92e5-00bafac1326b">
2. Does your implementation support generative PC networks? In particular, I'd like to make forward predictions (i.e., clamp the class label and generate data) _and_ backward predictions (i.e., clamp a data point and infer the class). In other words, can you run the PyTorch PC model in reverse and if so, how? For instance, suppose my inputs are data labels and outputs are data points (i.e., that's how I define the PyTorch model). And now I want to fix a data point and predict its class, how would I do that? Fix the target to the data point and then set [`is_optimize_inputs=True`](https://github.com/Bogacz-Group/PredictiveCoding/blob/334a87eeb6c6cad5197de89453c2c0a9ff43a18a/predictive_coding/pc_trainer.py#L507) and read out the inferred activity?
3. Do you have any examples of unsupervised representation learning using PC for MNIST? Is it as simple as removing `loss_fn` in the call to `trainer.train_on_batch`? Or would it be better to keep a squared error loss function and set the target vector to be the all 0s vector. I'm imagining this would correspond to a 0-mean Gaussian prior. Or is there a better way to do this?

I'm somewhat new to predictive coding so please forgive me if I'm misunderstanding anything!

Thanks for the illuminating codebase!

**Comment by ToddMorrill** [comment]
I've been meaning to share the unsupervised model that I built with this codebase. Here it is in case you find it valuable as another example for the repo. You could imagine adding sparsity priors on the activity and then showcasing this as the classic Rao and Ballard formulation. Thanks for the nice repo.

https://toddmorrill.github.io/machine%20learning/unsupervised-predictive-coding/
